File: app/src/main/java/com/hiddenramblings/tagmo/eightbit/bluetooth/GattArray.java
package com.hiddenramblings.tagmo.eightbit.bluetooth;

import android.bluetooth.BluetoothGattCharacteristic;
import android.os.Handler;
import android.os.Looper;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class GattArray {

    public static List<byte[]> byteToPortions(byte[] largeByteArray, int sizePerPortion) {
        List<byte[]> byteArrayPortions = new ArrayList<>();
        int offset = 0;
        while (offset < largeByteArray.length) {
            byte[] portion = Arrays.copyOfRange(largeByteArray, offset, offset + sizePerPortion);
            offset += sizePerPortion;
            byteArrayPortions.add(portion);
        }
        return byteArrayPortions;
    }

    public static List<String> stringToPortions(String largeString, int sizePerPortion) {
        List<String> stringPortions = new ArrayList<>();
        int size = largeString.length();
        if (size <= sizePerPortion) {
            stringPortions.add(largeString);
        } else {
            int index = 0;
            while (index < size) {
                stringPortions.add(largeString.substring(index,
                        Math.min(index + sizePerPortion, largeString.length())));
                index += sizePerPortion;
            }
        }
        return stringPortions;
    }

    public static String stringToUnicode(String s) {
        StringBuilder sb = new StringBuilder(s.length() * 3);
        for (char c : s.toCharArray()) {
            if (c < 256) {
                sb.append(c);
            } else {
                String strHex = Integer.toHexString(c);
                sb.append("\\u").append(strHex);
            }
        }
        return sb.toString();
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/bluetooth/GattArray.kt
/*
 * ====================================================================
 * Copyright (C) 2022 AbandonedCart @ TagMo
 * ====================================================================
 */
package com.hiddenramblings.tagmo.bluetooth

import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import com.hiddenramblings.tagmo.nfctech.Foomiibo
import com.hiddenramblings.tagmo.nfctech.NfcByte

object GattArray {
    fun BluetoothGattCharacteristic.hasProperty(property: Int): Boolean {
        return this.properties and property != 0
    }

    fun BluetoothGatt.getCharacteristicByProperty(property: Int): BluetoothGattCharacteristic? {
       return services.find { service ->
           service.characteristics.any { it.hasProperty(property) }
       }?.characteristics?.find { it.hasProperty(property) }
    }

    fun generateBlank(): ByteArray {
        val blankData = ByteArray(NfcByte.TAG_FULL_SIZE)
        Foomiibo.generateRandomUID().plus(byteArrayOf(
                0x48, 0x00, 0x00, 0xE1.toByte(), 0x10, 0x3E, 0x00, 0x03, 0x00, 0xFE.toByte()
        )).copyInto(blankData, 0x00)
        byteArrayOf(
                0xBD.toByte(), 0x04, 0x00, 0x00, 0xFF.toByte(), 0x00, 0x05
        ).copyInto(blankData, 0x20B)
        return blankData
    }

    @JvmStatic
    fun ByteArray.toPortions(sizePerPortion: Int): ArrayList<ByteArray> {
        val byteArrayPortions: ArrayList<ByteArray> = arrayListOf()
        this.asIterable().chunked(sizePerPortion).forEach {
            byteArrayPortions.add(it.toByteArray())
        }
        return byteArrayPortions
    }

    @JvmStatic
    fun String.toUnicode(): String {
        val sb = StringBuilder(this.length * 3)
        for (c in this.toCharArray()) {
            if (c.code < 256) {
                sb.append(c)
            } else {
                val strHex = Integer.toHexString(c.code)
                sb.append("\\u").append(strHex)
            }
        }
        return sb.toString()
    }

    fun String.toMilliseconds(): Long {
        return this.toLong() / 10
    }
}